### PR TITLE
Altering the translate and optimize signatures

### PR DESCRIPTION
**Pull request by kriskowal**
They all just mutate the module arguments and have the option of
returning promises.

Also grabs new versions of dependencies and moves onto ^ dependencies.

### CODE REVIEW

File: boot.js [hunk 1]
```diff
@@ -96,1994 +96,304 @@ function boot(preloaded, params) {
 
 }
 
-}],["asap","asap",{},function (require, exports, module){
+}],["asap","browser-asap",{"./raw":2},function (require, exports, module){
 
-// asap asap
```
Comment (kriskowal) [inline boot.js:101]: New version of Q upgrades ASAP and gets rid of its dependency on Collections in favor of depending directly on weak-map and pop-iterate.

File: boot.js [hunk 2]
```diff
@@ -3112,11 +1429,14 @@ function configurePackage(location, description, parent) {
...
-                redirect: Identifier.normalize(Identifier.resolve(redirects[name], "")),
-                location: URL.resolve(location, name)
+        Object.keys(redirects).forEach(function (source) {
+            var target = redirects[source];
+            source = Identifier.normalize(Identifier.resolve(source, ""));
+            target = Identifier.normalize(Identifier.resolve(target, ""));
+            modules[source] = {
+                id: source,
+                redirect: target,
+                location: URL.resolve(location, target)
```
Comment (kriskowal) [inline boot.js:1439]: This fixes ASAP now that ASAP uses redirects that are compatible with Browserify. The keys of browserify redirects are relative for local and top-level for for other packages or Node.js internals. Mr has a single namespace for both.

File: test/translator/text-translator.js [hunk 3]
```diff
@@ -1,3 +1,3 @@
-module.exports = function (text) {
-    return "module.exports = " + JSON.stringify(text) + ";";
+module.exports = function (module) {
+    module.text = "module.exports = " + JSON.stringify(module.text) + ";";
 };
```
Comment (kriskowal) [inline test/translator/text-translator.js:3]: These reflect the substantiative changes. Translators and optimizers accept the module object, modify it, and may return a promise. This makes the signatures of translators, optimizers, and compilers more uniform.

**Comment by kriskowal** [comment]
cc @stuk.

**Comment by Stuk** [comment]
Changes look good to me. `npm test` is failing though:

```
$ npm test

> mr@2.0.4 test /Users/stuart/Documents/montagejs/mr
> npm run build && npm run test:node && npm run test:phantom && npm run lint


> mr@2.0.4 build /Users/stuart/Documents/montagejs/mr
> node bin/mrs.js boot/require.js -e '().done()' > boot.js

/Users/stuart/Documents/montagejs/mr/node_modules/q/q.js:876
    throw error;
          ^
Error: Error: ENOENT, open '/Users/stuart/Documents/montagejs/mr/node_modules/jasminum/node_modules/qs/package.json'
    at /Users/stuart/Documents/montagejs/mr/require.js:52:29
    at fs.js:224:20
    at Object.oncomplete (fs.js:93:15)
```

**Comment by kriskowal** [comment]
@stuk, run npm install.

**Comment by Stuk** [comment]
I have, still get the same errors.

**Pull request merged**